toolkit/tools/imagegen/diskutils/verity.go

```diff
@@ -0,0 +1,316 @@
+	if err != nil {
+		return
+	}
+
+	for _, file := range verityFiles {
+		filePath := filepath.Join(verityWorkingDirectory, file.Name())
+		logger.Log.Debugf("Adding '%s' to initramfs", filePath)
+		// Place each file in the root of the initramfs
+		err = initramfs.AddFileToInitramfs(filePath, file.Name())
+		if err != nil {
+			return fmt.Errorf("failed adding %s to initramfs: %w", filePath, err)
+		}
+	}
+
+	return
+}
+
+func (v *VerityDevice) createVerityDisk(verityDirectory string) (err error) {
+	const (
+		hexChars   = "0123456789abcdef"
+		saltLength = 64
+		hashAlg    = "sha256"
+	)
+	var (
+		verityFecArgs    []string
+		verityArgs       []string
+		verityVerifyArgs []string
+		fileBase         = filepath.Join(verityDirectory, v.MappedName)
+		rootHashPath     = fmt.Sprintf("%s.roothash", fileBase)
+		hashtreePath     = fmt.Sprintf("%s.hashtree", fileBase)
+		fecFilePath      = fmt.Sprintf("%s.fec", fileBase)
+	)
+
+	err = os.MkdirAll(verityDirectory, os.ModePerm)
+	if err != nil {
+		return
+	}
+
+	rootHashFile, err := os.Create(rootHashPath)
+	if err != nil {
+		return
+	}
+	defer rootHashFile.Close()
+
+	salt, err := randomization.RandomString(64, hexChars)
+	if err != nil {
+		return
+	}
+
+	if v.FecRoots > 0 {
+		verityFecArgs = []string{
+			fmt.Sprintf("--fec-device=%s", fecFilePath),
+			fmt.Sprintf("--fec-roots=%d", v.FecRoots),
+		}
+	}
+
+	verityArgs = []string{
+		"--salt",
+		salt,
+		"--hash",
+		hashAlg,
+		"--verbose",
+		"--debug",
+		"format",
+		v.MappedDevice,
+		hashtreePath,
+	}
+
+	logger.Log.Info("Generating a dm-verity read-only partition, this may take several minutes")
+	verityOutput, stderr, err := shell.Execute("veritysetup", append(verityFecArgs, verityArgs...)...)
+	if err != nil {
+		err = fmt.Errorf("Unable to create verity disk '%s': %w", stderr, err)
+		return
+	}
+
+	// Searches for a line like: "Root hash:      1234567890abcdefg..."
+	rootHashLine, stderr, err := shell.ExecuteWithStdin(verityOutput, "grep", "^Root hash:")
+	if err != nil {
+		err = fmt.Errorf("Unable to find root hash '%s': %w", stderr, err)
+		return
+	}
+	// Print the last entry in the line
+	// ORS (output record separator) inserts a newline normally, explicitly skip it
+	rootHash, stderr, err := shell.ExecuteWithStdin(rootHashLine, "awk", `BEGIN {ORS=""}; END{ print $NF }`)
+	if err != nil {
+		err = fmt.Errorf("Unable extract root hash '%s': %w", stderr, err)
+		return
+	}
+
+	logger.Log.Infof("Verity partition completed, root hash: '%s'", rootHash)
+	_, err = rootHashFile.WriteString(rootHash)
+	if err != nil {
+		return
+	}
+
+	//Verify the disk was created correctly:
+	verityVerifyArgs = []string{
+		"--verbose",
+		"--debug",
+		"verify",
+		v.MappedDevice,
+		hashtreePath,
+		rootHash,
+	}
+
+	logger.Log.Info("Verifying the verity partition")
+	verityOutput, stderr, err = shell.Execute("veritysetup", verityVerifyArgs...)
+	if err != nil {
+		logger.Log.Error(verityOutput)
```

Comment:
I suggest wrapping it in some context with `Errorf` to keep the message meaningful in the future